kernel/packages/voice-chat-codec/worker.ts

```diff
@@ -0,0 +1,157 @@
+
+declare const libopus: any
+
+self.LIBOPUS_WASM_URL = 'libopus.wasm'
+
+importScripts('libopus.wasm.js')
+
+type CodecWorklet = {
+  working: boolean
+  lastWorkTime: number
+  destroy: () => any
+}
+
+type EncoderWorklet = {
+  encoder: any
+} & CodecWorklet
+
+type DecoderWorklet = {
+  decoder: any
+} & CodecWorklet
+
+function getSampleRate(e: MessageEvent) {
+  return e.data.sampleRate ? e.data.sampleRate : 24000
+}
+
+const encoderWorklets: Record<string, EncoderWorklet> = {}
+const decoderWorklets: Record<string, DecoderWorklet> = {}
+
+function startWorklet<T extends CodecWorklet, O extends Uint8Array | Float32Array>(
+  streamId: string,
+  worklet: T,
+  outputFunction: (worklet: T) => O,
+  messageBuilder: (output: O, streamId: string) => VoiceChatWorkerResponse
+) {
+  worklet.working = true
+
+  function doWork() {
+    worklet.lastWorkTime = Date.now()
+
+    let output = outputFunction(worklet)
+
+    if (output) {
+      if (output instanceof Uint8Array) {
+        output = Uint8Array.from(output) as O
+      } else {
+        output = Float32Array.from(output) as O
+      }
+
+      postMessage(messageBuilder(output, streamId), [output.buffer])
+      setTimeout(doWork, 0)
+    } else {
+      worklet.working = false
+    }
+  }
+
+  setTimeout(doWork, 0)
+}
+
+//Encoder(channels, samplerate, bitrate, frame_size, voice_optimization)
+
+onmessage = function (e) {
+  if (e.data.topic === RequestTopic.ENCODE) {
+    const sampleRate = getSampleRate(e)
+    const encoderWorklet = (encoderWorklets[e.data.streamId] = encoderWorklets[e.data.streamId] || {
+      working: false,
+      encoder: new libopus.Encoder(1, sampleRate, 24000, 20, true),
+      lastWorkTime: Date.now(),
+      destroy: function () {
+        this.encoder.destroy()
+      }
+    })
+
+    const samples = toInt16Samples(resampleIfNecessary(e.data.samples, e.data.sampleRate, e.data.inputSampleRate))
+
+    encoderWorklet.encoder.input(samples)
+
+    if (!encoderWorklet.working) {
+      startWorklet(
+        e.data.streamId,
+        encoderWorklet,
+        (worklet) => worklet.encoder.output(),
+        (output, streamId) => ({ topic: ResponseTopic.ENCODE, streamId: streamId, encoded: output })
+      )
+    }
+  }
+
+  if (e.data.topic === RequestTopic.DECODE) {
+    const sampleRate = getSampleRate(e)
+    const decoderWorklet = (decoderWorklets[e.data.streamId] = decoderWorklets[e.data.streamId] || {
+      working: false,
+      decoder: new libopus.Decoder(1, sampleRate),
+      lastWorkTime: Date.now(),
+      destroy: function () {
+        this.decoder.destroy()
+      }
+    })
+
+    decoderWorklet.decoder.input(e.data.encoded)
+
+    if (!decoderWorklet.working) {
+      startWorklet(
+        e.data.streamId,
+        decoderWorklet,
+        (worklet) => worklet.decoder.output(),
+        (output, streamId) => ({
+          topic: ResponseTopic.DECODE,
+          streamId,
+          samples: toFloat32Samples(output)
+        })
+      )
+    }
```

Comment:
It wouldn't hurt to extract the `encoder` and `decoder` bits into separate functions, so the code becomes a bit more readable.

Reply:
Done!